libshaderc/src/shaderc.cc

```diff
@@ -85,6 +96,60 @@ struct {
 
 std::mutex compile_mutex;  // Guards shaderc_compile_*.
 
+// A wrapper functor class to be used as stage deducer for libshaderc_util
+// Compile() interface. When the given shader kind is one of the default shader
+// kinds, this functor will be called if #pragma is not found in the source
+// code. And it returns the corresponding shader stage. When the shader kind is
+// a forced shader kind, this functor won't be called and it simply returns
+// EShLangCount to make the syntax correct. When the shader kind is set to
+// shaderc_glsl_deduce_from_pragma, this functor also returns EShLangCount, but
+// the compiler should emit error if #pragma annotation is not found in this
+// case.
+class StageDeducer {
+ public:
+  StageDeducer(shaderc_shader_kind kind) : kind_(kind){};
+  EShLanguage operator()(std::ostream* error_stream,
+                         const shaderc_util::string_piece& error_tag) const {
+    EShLanguage stage = GetDefaultStage(kind_);
+    if (stage == EShLangCount) {
+      *error_stream << "libshaderc error: '" << error_tag
+                    << "' shader kind not determined" << std::endl;
+    }
+    return stage;
+  };
+  StageDeducer() : kind_(shaderc_glsl_deduce_from_pragma){};
```

Comment:
I think the style guide allows a default argument in this case.  It might improve readability.
Coder's choice.